File: golang/hashmap/valid_anagram.go
package main

func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	charCount := make(map[rune]int)

	for _, char := range s {
		charCount[char]++
	}

	for _, char := range t {
		if _, exists := charCount[char]; exists {
			charCount[char]--

			if charCount[char] < 0 {
				return false
			}
		} else {
			return false
		}
	}

	for _, count := range charCount {
		if count != 0 {
			return false
		}
	}

	return true
}

File: python/hashmap/valid_anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        char_count = {}

        for char in s:
            if char in char_count:
                char_count[char] += 1
            else:
                char_count[char] = 1

        for char in t:
            if char in char_count:
                char_count[char] -= 1

                if char_count[char] < 0:
                    return False

            else:
                return False

        return all(value == 0 for value in char_count.values())
